src/neptune_scale/sync/operations_repository.py

```diff
@@ -0,0 +1,261 @@
+      - fork_step REAL: Fork step (optional)
+    """
+
+    def __init__(self, db_path: str) -> None:
+        self._db_path = db_path
+        self._lock = threading.RLock()
+        self._connection: Optional[sqlite3.Connection] = None
+
+    def init_db(self) -> None:
+        os.makedirs(os.path.dirname(os.path.abspath(self._db_path)), exist_ok=True)
+        with self._get_connection() as conn:  # type: ignore
+            conn.execute(
+                """
+                CREATE TABLE IF NOT EXISTS run_operations (
+                    sequence_id INTEGER PRIMARY KEY AUTOINCREMENT,
+                    timestamp INTEGER NOT NULL,
+                    operation_type INTEGER NOT NULL,
+                    operation BLOB NOT NULL,
+                    metadata_size INTEGER NOT NULL
+                )
+                """
+            )
+            conn.execute(
+                """
+                CREATE TABLE IF NOT EXISTS metadata (
+                    id INTEGER PRIMARY KEY AUTOINCREMENT,
+                    version TEXT NOT NULL,
+                    project TEXT NOT NULL,
+                    run_id TEXT NOT NULL,
+                    parent TEXT,
+                    fork_step REAL
+                )"""
+            )
+
+    def save_update_run_snapshots(self, ops: list[UpdateRunSnapshot]) -> SequenceId:
+        current_time = int(time.time() * 1000)  # milliseconds timestamp
+        params = []
+
+        for update in ops:
+            serialized_operation = update.SerializeToString()
+            metadata_size = len(serialized_operation)
+            params.append((current_time, OperationType.UPDATE_SNAPSHOT, serialized_operation, metadata_size))
+
+        with self._get_connection() as conn:  # type: ignore
+            cursor = conn.cursor()
+
+            cursor.executemany(
+                "INSERT INTO run_operations (timestamp, operation_type, operation, metadata_size) VALUES (?, ?, ?, ?)",
+                params,
+            )
+            cursor.execute("SELECT last_insert_rowid()")
+            last_insert_rowid: int = cursor.fetchone()[0]
+
+            return SequenceId(last_insert_rowid)
+
+    def save_create_run(self, run: CreateRun) -> SequenceId:
+        with self._get_connection() as conn:  # type: ignore
+            cursor = conn.cursor()
+
+            current_time = int(time.time() * 1000)  # milliseconds timestamp
+            serialized_operation = run.SerializeToString()
+            metadata_size = len(serialized_operation)
+
+            cursor.execute(
+                "INSERT INTO run_operations (timestamp, operation_type, operation, metadata_size) VALUES (?, ?, ?, ?)",
+                (current_time, OperationType.CREATE_RUN, serialized_operation, metadata_size),
+            )
+            return SequenceId(cursor.lastrowid)  # type: ignore
+
+    def get_operations(self, up_to_bytes: int, window_function_limit: int = 10_000) -> list[Operation]:
```

Comment:
I don't like this additional `window_function_limit` parameter here. It's clearly associated with this particular implementation, has nothing to do with the semantics of this function.

Additionally, I suspect this query can OOM with long enough rows. `running_size` is not lazy.

Would it maybe make sense to:
1. Read seq_id, size in a loop in chunks of 10k
2. Once we know we have enough rows, we read the actual data
?

In most cases that's 2 queries, both trivial, mem-efficient.

Reply:
https://github.com/neptune-ai/neptune-client-scale/blob/dd12d23bb9ef65449143cea6d28f014828fbcb58/src/neptune_scale/sync/operations_repository.py#L142-L183